pytd/writer.py

```diff
@@ -265,14 +278,63 @@ def write_dataframe(self, dataframe, table, if_exists, fmt="csv"):
             - msgpack
                 Convert to temporary msgpack.gz file. Fast option but might be
                 unstable because of skipping ``tdclient.API._prepare_file()``
+
+        keep_list : boolean, default: False
+            If this argument is True, keep list or numpy.ndarray column as list, which
+            will be converted array<T> on Treasure Data table.
+            Each type of element of list will be converted by
+            ``numpy.array(your_list).tolist()``.
+
+            Examples
+            ---------
+
+            A dataframe containing list will be treated array<T> in TD.
+
+            >>> import pytd
+            >>> import numpy as np
+            >>> import pandas as pd
+            >>> df = pd.DataFrame(
+            ...     {
+            ...         "A": [[1, 2, 3], [2, 3, 4]],
+            ...         "B": [[0, None, 2], [2, 3, 4]],
+            ...     }
+            ... )
+            >>> client = pytd.Client()
+            >>> table = pytd.table.Table(client, "mydb", "test")
+            >>> writer = pytd.writer.BulkImportWriter()
+            >>> writer.write_dataframe(df, table, if_exists="overwrite", keep_list=True)
+
+            In this case, the type of columns will be:
+            ``{"A": array<int>, "b": array<string>}``
+
+            If you want to set the type after ingestion, you need to run
+            ``tdclient.Client.update_schema`` like:
+
+            >>> client.api_client.update_schema(
+            ...     "mydb",
+            ...     "test",
+            ...     [["a", "array<long>", "a"], ["b", "array<int>", "b"],
+            ... )
+
+            Note that ``numpy.nan`` will be converted as a string value as ``"NaN"``.
+            Also, numpy converts integer array including np.nan into float array because
+            ``np.nan`` is a Floating Point Special Value. See also:
+            https://docs.scipy.org/doc/numpy-1.13.0/user/misc.html#ieee-754-floating-point-special-values
+
+            Or, you can use :func:`Client.load_table_from_dataframe` function as well.
+            >>> client.load_table_from_dataframe(df, "bulk_import", keep_list=True)
         """
         if self.closed:
             raise RuntimeError("this writer is already closed and no longer available")
 
         if "time" not in dataframe.columns:  # need time column for bulk import
             dataframe["time"] = int(time.time())
 
-        _cast_dtypes(dataframe)
+        # We enforce using "msgpack" format for list since CSV can't handle list.
+        if keep_list:
+            fmt = "msgpack"
```

Comment:
Is it okay to forcibly modify the format rather than raising an exception?

Anyways, can you mention this constraint in the docstring of `fmt` argument?

Reply:
I think it's ok to overwrite the fmt variable for convenience.

Reply:
Modified docstring 8c60124fc09547a249dca51118a42766828dbac6